server/src/main/java/io/crate/metadata/TransactionContext.java

```diff
@@ -29,20 +29,28 @@
 public interface TransactionContext {
 
     static TransactionContext of(SessionSettings sessionSettings) {
-        return new StaticTransactionContext(sessionSettings);
+        return of(sessionSettings, null);
+    }
+
+    static TransactionContext of(SessionSettings sessionSettings, Functions functions) {
+        return new StaticTransactionContext(sessionSettings, functions);
     }
 
     Instant currentInstant();
 
     SessionSettings sessionSettings();
 
+    Functions functions();
```

Comment:
I'm not sure if the functions fit into the `TransactionContext` - as the name implies it should contain state that is bound to the lifecycle of a transaction.

The Functions have a different lifecycle or scope. Maybe we should instead extend the `evaluate` method to take the `Functions` directly - or introduce some kind of `NodeContext`. Afaik we'll have other functions later that require access to the `Schemas`

@kovrus @seut what your take here?

Reply:
I am not sure about extending the `Scalar#evaluate` method with the `Functions` argument, that would result in something like: Register a function -> AbstractFunctionModule -> inject into Functions -> function evaluate(..., Functions) which seems kind of cyclic and also, as you mention, won't be helpful when implementing the functions that need access to `Schemas`. I'd probably prefer some `NodeContext` that would have `Functions` injected and later `Schemas`.

Reply:
@mfussenegger I have a different view.

`TransactionState` would clearly imply _"it should contain state that is bound to the lifecycle of a transaction"_.
`TransactionContext`, on the other hand, implies "the circumstances that form the setting for an event, statement, or idea, and in terms of which it can be fully understood" ([citation](https://www.google.com/search?q=context&oq=context&aqs=chrome..69i57j0l4j69i60l3.1581j0j9&sourceid=chrome&ie=UTF-8)).

In my mind, for the successful execution of one function (this one at least, for now), its context requires access to the other Functions, within a given transaction. Schemas can be derived from the `functions` map.

Please advise how to proceed.

Reply:
I would need some feedback to continue. 

- NodeContext contains a reference to Functions and is passed to Scalar.(evaluate, evaluateIfLiterals) and FunctionImplementation.normalizeSymbol (because of Scalar.evaluateIfLiterals' signature).
- From there, I notise Analyzer's constructor creates all analyzers, and these could take NodeContext as a parameter, to have access to Functions. I could add Schemas to NodeContext, now or at a later PR.

Reply:
left some comments. I think it would be good to also get the input from @seut because the interface change touches a lot of components

Reply:
As I understand this proposal, `NodeContext` will superseed the usage of `Functions` as it will hold a reference to it. So I think, `NodeContext` should be used in a same ways than `Functions`. 
In detail:
 - `NodeContext` should be a singleton
 - most (if not any) usages of `Functions` should be replaced by `NodeContext.getFunctions()`
 -  I'd prefer to wrap`Schemas` into `NodeContext` later on when required to lower the scope of this PR

Beside of that I agree on this proposal (like introducing `NodeContext` at all).